### PR TITLE
Added location to R4 encounter write

### PR DESCRIPTION
**Pull request by ahamdany**
Description
----
Added location to R4 encounter write
Screen Shot 2022-02-15 at 4 14 38 PM

Screen Shot 2022-02-15 at 4 12 29 PM

Screen Shot 2022-02-15 at 4 11 48 PM

PR Checklist
----
- [ ] Screenshot(s) of changes attached before changes merged.
- [ ] Screenshot(s) of changes attached after changes merged and published.

### CODE REVIEW

File: lib/resources/example_json/r4_examples_encounter.rb [hunk 1]
```diff
@@ -1983,6 +1983,13 @@ module Resources
       'serviceProvider': {
         'reference': 'Organization/675844'
       },
+      "location": [
+          {
+            "location": {
+              "reference": "Location/633869"
```
Comment (alexbeckner1) [inline lib/resources/example_json/r4_examples_encounter.rb:1989]: can we use  a valid location reference from c1941 in our examples? I think 21250409 is valid.

File: lib/resources/r4/encounter.yaml [hunk 2]
```diff
@@ -528,6 +528,22 @@ fields:
       }
     }
 
+- name: location
+  required: 'No'
```
Comment (alexbeckner1) [inline lib/resources/r4/encounter.yaml:532]: Should required be 'Yes' since either serviceProvider or location must be provided?

File: lib/resources/r4/encounter.yaml [hunk 3]
```diff
@@ -528,6 +528,22 @@ fields:
...
+  type: Reference (Location)
+  description: The location where the encounter takes place.
+  action: create
+  example: |
+    {
+      "location": [
+        {
+          "location": {
+            "reference": "Location/633869"
```
Comment (alexbeckner1) [inline lib/resources/r4/encounter.yaml:541]: same as above

File: content/millennium/r4/base/management/encounter.md [hunk 4]
```diff
@@ -225,6 +227,12 @@ Create an individual Encounter.
 
     POST /Encounter
 
+_Implementation Notes_
+
+* Only one [Encounter.location.location] reference is permitted.
+* [Encounter.serviceProvider] and [Encounter.location.location] are mutually exclusive. Only one or the other is permitted.
```
Comment (alexbeckner1) [inline content/millennium/r4/base/management/encounter.md:233]: How about rewording this to something like "[Encounter.location.location] cannot be provided when [Encounter.serviceProvider] is valued." We could also have a similar note in the implementation notes for serviceProvider field.

File: content/millennium/r4/base/management/encounter.md [hunk 5]
```diff
@@ -225,6 +227,12 @@ Create an individual Encounter.
 
     POST /Encounter
 
+_Implementation Notes_
```
Comment (alexbeckner1) [inline content/millennium/r4/base/management/encounter.md:230]: Let's move these notes down to the Location specific Notes in the Body fields section.

File: content/millennium/r4/base/management/encounter.md [hunk 6]
```diff
@@ -183,6 +183,8 @@ _Implementation Notes_
 
 * The [Encounter.hospitalization.destination] will be returned as a reference to a [contained] location resource.
 * The [Encounter.location.location] may be returned as a reference to a [contained] location resource.
+* Only one [Encounter.location.location] reference is permitted.
```
Comment (alexbeckner1) [inline content/millennium/r4/base/management/encounter.md:186]: I don't think we need these implementation notes related to the Retrieve by id operation.

File: lib/resources/r4/encounter.yaml [hunk 7]
```diff
@@ -528,6 +528,22 @@ fields:
...
+          }
+        }
+      ]
+    }
+
```
Comment (alexbeckner1) [inline lib/resources/r4/encounter.yaml:546]: This is where body field notes should go

File: lib/resources/example_json/r4_examples_encounter.rb [hunk 8]
```diff
@@ -1983,6 +1983,13 @@ module Resources
       'serviceProvider': {
         'reference': 'Organization/675844'
       },
+      "location": [
```
Comment (Coffeevee) [inline lib/resources/example_json/r4_examples_encounter.rb:1986]: nit - let's use single quotes on this example

File: content/millennium/r4/base/management/encounter.md (hunk 6 above)
Comment (ahamdany) [reply content/millennium/r4/base/management/encounter.md]: whoops that shouldn't be there, removed https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-3d460d456f4275cf030a4b8eb00128a8f017087dd7a2918741e96202c45e6c52L186

File: content/millennium/r4/base/management/encounter.md (hunk 5 above)
Comment (ahamdany) [reply content/millennium/r4/base/management/encounter.md]: moved https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-3d460d456f4275cf030a4b8eb00128a8f017087dd7a2918741e96202c45e6c52L230 and https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-ea755c3df7ed8869a54a274cedf2e40e119b2f16a6e3ed8bc8ae211e0e87389cR550

File: content/millennium/r4/base/management/encounter.md (hunk 4 above)
Comment (ahamdany) [reply content/millennium/r4/base/management/encounter.md]: changed wording https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-ea755c3df7ed8869a54a274cedf2e40e119b2f16a6e3ed8bc8ae211e0e87389cR553 and also added to serviceProvider https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-ea755c3df7ed8869a54a274cedf2e40e119b2f16a6e3ed8bc8ae211e0e87389cR532

File: lib/resources/r4/encounter.yaml (hunk 3 above)
Comment (ahamdany) [reply lib/resources/r4/encounter.yaml]: [done](https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-ea755c3df7ed8869a54a274cedf2e40e119b2f16a6e3ed8bc8ae211e0e87389cR545) and https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-bec48e0a1467ab0b687a5ea45ebd8dcdf17aabfbf446f8e0aa5b0a6b7aa6ba4eR1989

File: lib/resources/r4/encounter.yaml (hunk 2 above)
Comment (ahamdany) [reply lib/resources/r4/encounter.yaml]: changed it to yes https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-ea755c3df7ed8869a54a274cedf2e40e119b2f16a6e3ed8bc8ae211e0e87389cR536 the notes can do the job for the  mutual exclusive part.

File: lib/resources/r4/encounter.yaml (hunk 7 above)
Comment (ahamdany) [reply lib/resources/r4/encounter.yaml:555]: yup added here https://github.com/cerner/fhir.cerner.com/pull/803/commits/1d0f2634bd131127317a63f5458e2f27fc579829#diff-ea755c3df7ed8869a54a274cedf2e40e119b2f16a6e3ed8bc8ae211e0e87389cL545

File: lib/resources/example_json/r4_examples_encounter.rb (hunk 8 above)
Comment (Coffeevee) [reply lib/resources/example_json/r4_examples_encounter.rb]: left a small follow-up comment on the commit

File: lib/resources/r4/encounter.yaml [hunk 9]
```diff
@@ -527,6 +527,31 @@ fields:
         "reference": "Organization/619848"
       }
     }
+  note: >
+    <ul>
+      <li>serviceProvider cannot be provided when Encounter.location.location is present.</li>
```
Comment (exopsRamba) [inline lib/resources/r4/encounter.yaml:532]: nit: let's put field names in ```code```.

Comment (exopsRamba) [reply lib/resources/r4/encounter.yaml:532]: Also we should be consistent with our naming. Either both should be 'serviceProvider' and 'location' or both should be 'Encounter.serviceProvider' and 'Encounter.location'.

File: lib/resources/r4/encounter.yaml [hunk 10]
```diff
@@ -527,6 +527,31 @@ fields:
...
+          "location": {
+            "reference": "Location/21250409"
+          }
+        }
+      ]
+    }
+  note: >
+    <ul>
+      <li>Only one location reference is permitted.</li>
+      <li>location cannot be provided when Encounter.serviceProvider is present.</li>
```
Comment (exopsRamba) [inline lib/resources/r4/encounter.yaml:553]: nit: location in ```code```.

File: lib/resources/example_json/r4_examples_encounter.rb [hunk 11]
```diff
@@ -1983,6 +1983,13 @@ module Resources
       'serviceProvider': {
         'reference': 'Organization/675844'
       },
+      'location': [
+          {
+            'location': {
+              'reference': "Location/21250409"
+            }
+          }
+        ],
```
[the remark this answers is not in the document]
Comment (ahamdany) [reply lib/resources/example_json/r4_examples_encounter.rb:1992]: I thought about that but wasnt sure if we should keep it or not, since we added the mutual exclusiveness in the notes, I decided to keep it. Let me know if its better removed.

File: lib/resources/r4/encounter.yaml (hunk 9 above)
Comment (ahamdany) [reply lib/resources/r4/encounter.yaml]: [done](https://github.com/cerner/fhir.cerner.com/pull/803/commits/62e556c0c21d59fccf0d1ab62663a549c0045d74#diff-ea755c3df7ed8869a54a274cedf2e40e119b2f16a6e3ed8bc8ae211e0e87389cR532)

File: lib/resources/r4/encounter.yaml (hunk 10 above)
Comment (ahamdany) [reply lib/resources/r4/encounter.yaml]: https://github.com/cerner/fhir.cerner.com/pull/803/commits/62e556c0c21d59fccf0d1ab62663a549c0045d74#diff-ea755c3df7ed8869a54a274cedf2e40e119b2f16a6e3ed8bc8ae211e0e87389cR552

**Review by Coffeevee — APPROVED** [review]

File: lib/resources/example_json/r4_examples_encounter.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (ahamdany) [reply lib/resources/example_json/r4_examples_encounter.rb]: removed https://github.com/cerner/fhir.cerner.com/pull/803/commits/d035a76b31bd6d684721dcc169256e92e8f738ca

**Review by alexbeckner1 — APPROVED** [review]

**Review by ramya1096 — APPROVED** [review]
